proposals/record-structs.md

```diff
@@ -0,0 +1,303 @@
+
+A record struct is not permitted to declare a parameterless primary constructor.
+
+Instance field declarations for a record struct are permitted to include variable initializers when there is a primary constructor. 
+At runtime the primary constructor executes the instance initializers appearing in the record-struct-body.
+
+If a record struct has a primary constructor, any user-defined constructor, except "copy constructor" must have an
+explicit `this` constructor initializer.
+
+Parameters of the primary constructor as well as members of the record struct are in scope within initializers of instance fields or properties. 
+Instance members would be an error in these locations (similar to how instance members are in scope in regular constructor initializers
+today, but an error to use), but the parameters of the primary constructor would be in scope and useable and
+would shadow members. Static members would also be useable.
+
+A warning is produced if a parameter of the primary constructor is not read.
+
+The definite assigment rules for struct instance constructors apply to the primary constructor of record structs. For instance, the following
+is an error:
+
+```csharp
+record struct Pos(int X) // definite assignment error in primary constructor
+{
+    private int x;
+    public int X { get { return x; } set { x = value; } } = X;
+}
+```
+
+### Properties
+
+For each record struct parameter of a record struct declaration there is a corresponding public property
+member whose name and type are taken from the value parameter declaration.
+
+For a record struct:
+
+* A public `get` and `set` auto-property is created.
+  An inherited `abstract` property with matching type is overridden.
+  It is an error if the inherited property does not have `public` overridable `get` and `set` accessors.
+  The auto-property is initialized to the value of the corresponding primary constructor parameter.
+  Attributes can be applied to the synthesized auto-property and its backing field by using `property:` or `field:`
+  targets for attributes syntactically applied to the corresponding record struct parameter.  
+
+### Deconstruct
+
+A positional record struct with at least one parameter synthesizes a public void-returning instance method called Deconstruct with an out
+parameter declaration for each parameter of the primary constructor declaration. Each parameter
+of the Deconstruct method has the same type as the corresponding parameter of the primary
+constructor declaration. The body of the method assigns each parameter of the Deconstruct method
+to the value from an instance member access to a member of the same name.
+The method can be declared explicitly. It is an error if the explicit declaration does not match
+the expected signature or accessibility, or is static.
+
+# Allow `with` expression on structs
+
+It is now valid for the receiver in a `with` expression to have a struct type.
```

Comment:
perhaps good to mention that this is not restricted to just `record struct` types.

Reply:
Meh. `struct type` seems clear enough (the rest of the spec says `record struct`).
Also, I don't have a good term for non-record structs. "Regular" structs, "non-record" structs, ...?

---
In reply to: [553565377](https://github.com/dotnet/csharplang/pull/4307#discussion_r553565377) [](ancestors = 553565377)